### PR TITLE
_WD_UpdateDriver - revised

### PR DESCRIPTION
**Pull request by mlipok**
Today I got problems with _WD_UpdateDriver() and I do not know where the issue was.
So I do some checking and thus my update.
There was redundant checking for @error in 
```
		$sDriverVersion = _WD_GetWebDriverVersion($sInstallDir, $sDriverEXE)
		If @error Then $iErr = @error
```
It was redundant because of:
```
		If $iErr = $_WD_ERROR_Success Then
			; When $bForce parameter equals Null, then return True if newer driver is available

```

and there was also lack of check for the case of `$sDriverVersion = 'None'` and thus when I have'nt already downloaded any `chromedriver.exe` or `geckodriver.exe` or `msedgedriver.exe` then this function never started downloading a newer version.
So I added them here:
`ElseIf $sDriverLatest > $sDriverVersion Or $bForce Or $sDriverVersion = 'None' Then
`

I also added some additional __WD_ConsoleWrite()

### CODE REVIEW

File: wd_helper.au3
Comment (Danp2) [inline wd_helper.au3:1392]: I'd have to go back and check past discussions, but I seem to recall that I previously told others that they needed to set $bForce = True in the case where the webdriver didn't previously exist.

**Review by Danp2 — CHANGES_REQUESTED** [review]

File: wd_helper.au3
Comment (Danp2) [reply wd_helper.au3:1391]: Found it. See discussion [here](https://www.autoitscript.com/forum/topic/201106-webdriver-udf-help-support-ii/?do=findComment&comment=1467771)

**Comment by mlipok** [comment]
Sorry for all my today comments to this PR.
I just notice that they was in PENDING state for last few days.
I just submit them now.

File: wd_helper.au3
Comment (Danp2) [reply wd_helper.au3:1329]: Good catch. It looks like I introduced the issue in 21d7269e620649a3ed87ee73b3edbbeec86c4d7f.

File: wd_helper.au3
Comment (Danp2) [reply wd_helper.au3:1391]: > so why `$bForce .... Default is False` ?
> Is this for security reason that Coders/Scripter must intentionaly download file ?

No. It's simply the way the function was designed to work.

> And the description is not accurate I mean misleading at least I was not aware about how to use them.

Can you be more specific? What would you change? We can certainly add more to the Remarks that clarify the functionality.

Comment (Danp2) [reply wd_helper.au3:1391]: > I do not want to update driver each run .
So how do you upgrade, but only if a newer version is available or no driver is available?

I see your point. In the current situation, you would need to call _WD_GetWebDriverVersion first, and then adjust your call to _WD_UpdateDriver as needed if the driver isn't present.

File: wd_helper.au3
Comment (mlipok) [reply wd_helper.au3:1329]: So what you propose here to solve this issue?
Is my fix good ?

File: wd_helper.au3
Comment (mlipok) [reply wd_helper.au3:1391]: > > And the description is not accurate I mean misleading at least I was not aware about how to use them.
> 
> Can you be more specific? What would you change? We can certainly add more to the Remarks that clarify the functionality.

Current description is:
`$bForce      - [optional] Force update? Default is False`

but it act like:
`$bForce      - [optional] Force download? Default is False`

because if you set True then it always download if you current local version is correct, which mean that you do not update but re download, which mean you forced to download regardles of any case. For me update mean only if there is lack of current version.

File: wd_helper.au3
Comment (Danp2) [reply wd_helper.au3:1329]: Not sure. I'll have to take a closer look at it when I have some time.

File: wd_helper.au3
Comment (Danp2) [reply wd_helper.au3:1391]: > Force update? vs Force download?

To me, they pretty much mean the same thing because you have to download the update in order to apply it, correct? In this scenario, how would you perform an update without downloading the file?

Also, the function name is _WD_UpdateDriver, not _WD_DownloadDriver. Thus the existing description works for me. <shrug>

**Comment by Danp2** [comment]
@mlipok Take a look at the latest commits and let me know if this looks better to you. I'm still not sure about the added check for "None". I'm wondering if 

- _WD_GetWebDriverVersion should return "0.0" instead of "None"
- Should we be converting the strings to numbers before comparing?

File: wd_helper.au3
Comment (mlipok) [reply wd_helper.au3:1329]: I check your commit ASAP

**Comment by Danp2** [comment]
@mlipok Can you explain why you merge main branch into PR? Is that really needed in this scenario?

File: wd_helper.au3
Comment (Danp2) [inline wd_helper.au3:1270]: I feel like we should also check for write access as well in case they do something stupid like passing in "c:\windows".

File: wd_helper.au3
Comment (mlipok) [reply wd_helper.au3]: Force update mean only when there is no dirver available or the driver is not up to date.
Force download .... regardless of the version you have - that is always.

Comment (mlipok) [reply wd_helper.au3]: Current version of this function works exactly in this way.

File: wd_helper.au3
Comment (mlipok) [inline wd_helper.au3:1348]: This is required due to the fact that `_WinAPI_GetBinaryType()` returns `$SCS_DOS_BINARY` and you can not determine if current version is 32/64bit because `msedgedriver.exe --version` and also `chromedriver.exe --version` returns the same information for `--version` for both 32 and 64 bit versions..

@Danp2  waiting for yours approval.
I think that this will be all in this PR

**Comment by mlipok** [comment]
> @mlipok Can you explain why you merge main branch into PR? Is that really needed in this scenario?

I supposed that this was required.

So I'm wondering now: If it was not required, when will it be required?

**Comment by mlipok** [comment]
btw. 
I think that this PR name `_WD_UpdateDriver - better @error handling` is not accurate.
Should it be changed? Do you have any suggestion?

File: wd_helper.au3
Comment (Danp2) [inline wd_helper.au3:1347]: Wouldn't it make sense to change this to `If $bDriverExists And FileGetSize($sInstallDir & $sDriverEXE) <> InetGetSize($sURLNewDriver, $INET_FORCERELOAD) Then $bForce = True`? No point in checking the file sizes if we know that the local file doesn't exist.

File: wd_helper.au3
Comment (mlipok) [reply wd_helper.au3]: I wonder why this is marked as `Outedated` as I resolve them before...
image

File: wd_helper.au3
Comment (mlipok) [reply wd_helper.au3]: Done

**Comment by mlipok** [comment]
> @mlipok Take a look at the latest commits and let me know if this looks better to you. I'm still not sure about the added check for "None". I'm wondering if
> 
> * _WD_GetWebDriverVersion should return "0.0" instead of "None"
> * Should we be converting the strings to numbers before comparing?

As far as I'm concerned, this is the last thing to check.

File: wd_helper.au3
Comment (Danp2) [reply wd_helper.au3]: In hindsight, this won't work because you are comparing the size of the local EXE to the remote ZIP, which will never match.

**Review by Danp2 — CHANGES_REQUESTED** [review]

File: wd_helper.au3
Comment (Danp2) [reply wd_helper.au3:1348]: > This is required due to the fact that _WinAPI_GetBinaryType() returns $SCS_DOS_BINARY

Are you sure about this? I got $SCS_32BIT_BINARY and $SCS_64BIT_BINARY when I tested it.

Comment (Danp2) [reply wd_helper.au3:1348]: Maybe a difference between Win7 and Win10?

Comment (mlipok) [reply wd_helper.au3]: using this code:

```AutoIt
#include <WinAPIFiles.au3>
ConsoleWrite("- " & @ScriptLineNumber & " " & _WinAPI_GetBinaryType("z:\Test\WebDriver\chromedriver.exe") & @CRLF)
ConsoleWrite("- " & @ScriptLineNumber & " " & _WinAPI_GetBinaryType("z:\Test\WebDriver\geckodriver.exe") & @CRLF)
ConsoleWrite("- " & @ScriptLineNumber & " " & _WinAPI_GetBinaryType("z:\Test\WebDriver\msedgedriver.exe") & @CRLF)

ConsoleWrite("- " & @ScriptLineNumber & " " & _WinAPI_GetBinaryType("z:\Test\WebDriver\1\chromedriver.exe") & @CRLF)
ConsoleWrite("- " & @ScriptLineNumber & " " & _WinAPI_GetBinaryType("z:\Test\WebDriver\1\geckodriver.exe") & @CRLF)
ConsoleWrite("- " & @ScriptLineNumber & " " & _WinAPI_GetBinaryType("z:\Test\WebDriver\1\msedgedriver.exe") & @CRLF)
Exit

```
I always get `1`

File: wd_helper.au3
Comment (mlipok) [reply wd_helper.au3]: reverted

File: wd_helper.au3
Comment (mlipok) [reply wd_helper.au3]: I use Windows 11 on my notebook.

Comment (mlipok) [reply wd_helper.au3]: Take a look here:
https://www.autoitscript.com/forum/topic/205553-webdriver-udf-help-support-iii/?do=findComment&comment=1495308

Comment (mlipok) [reply wd_helper.au3]: Done

**Comment by mlipok** [comment]
> > @mlipok Take a look at the latest commits and let me know if this looks better to you. I'm still not sure about the added check for "None". I'm wondering if
> > 
> > * _WD_GetWebDriverVersion should return "0.0" instead of "None"
> > * Should we be converting the strings to numbers before comparing?
> 
> As far as I'm concerned, this is the last thing to check.

_VersionCompare() ?

File: wd_helper.au3
Comment (Danp2) [inline wd_helper.au3:1299]: I think that we should consider either
- dropping the browser check here, or
- changing it to just check <> "chrome"

File: wd_helper.au3
Comment (Danp2) [inline wd_helper.au3:1478]: This is a script breaking change. I think some additional thought should go into this before proceeding.

Comment (Danp2) [reply wd_helper.au3:1478]: Also doesn't make sense to assign a numeric value to a string variable. 😛

File: wd_helper.au3
Comment (mlipok) [reply wd_helper.au3:1299]: > * dropping the browser check here, or

?

File: wd_helper.au3
Comment (mlipok) [reply wd_helper.au3:1478]: There was inconsistence with Function Header 
`Failure - "" (empty string) and sets @error to one of the following values:`
It was not true because at the top of the function it was declared as `$sDriverVersion = "None"`

File: wd_helper.au3
Comment (Danp2) [reply wd_helper.au3:1299]: > ?

The initial If...Endif could be viewed as unneeded as the function will still work without the "special handling" for Chrome not having a 64 bit version of the webdriver.

File: wd_helper.au3
Comment (mlipok) [reply wd_helper.au3:1478]: `_WD_GetWebDriverVersion()` was introduced not so long ago, and is used mainly in `_WD_Update()` 
Even if it is used by any coder/scripter should be checked with `_VersionCompare()` if it was'nt checked this way it was checked wrongly, so this is not ScriptBreaking Change, it is a fix.

Comment (Danp2) [reply wd_helper.au3:1478]: That fine. But it doesn't address my initial concerns posted above.

Comment (mlipok) [reply wd_helper.au3]: > Also doesn't make sense to assign a numeric value to a string variable. 😛

changed to `$sDriverVersion = "0"`

Comment (mlipok) [reply wd_helper.au3]: > I think some additional thought should go into this before proceeding.

I will wait for your thoughts ;)

Comment (mlipok) [reply wd_helper.au3]: Just checked:
even if we back to `Local $sDriverVersion = "None"` then `_WD_Update()` with using `_VersionCompare()` still works well.

File: wd_helper.au3
Comment (mlipok) [reply wd_helper.au3]: > * changing it to just check <> "chrome"

done

File: wd_helper.au3
Comment (Danp2) [reply wd_helper.au3]: I'm ok with changing it to "0", "0.00", or whatever. We just need to clearly document the change as potentially script breaking just in case someone is checking for "none".

File: wd_helper.au3
Comment (Danp2) [inline wd_helper.au3:1295]: Not sure why this is needed (there are 8 lines being output to the console, which is too much IMO), but there's something lacking here because it gets output even when no changes are being made. Perhaps you meant for this to go inside the `If $bDriverIs64Bit <> $bFlag64 Then` that follows.

File: wd_helper.au3
Comment (Danp2) [inline wd_helper.au3:1298]: Do we really want to delete the driver here? There's existing code that will delete the driver, so I think you could just set `$bForce = True` here.

**Review by Danp2 — CHANGES_REQUESTED** [review]

File: wd_helper.au3
Comment (mlipok) [reply wd_helper.au3]: Fixed.
And about the for 8 lines
Which one you think are handy, and which are useless ?

File: wd_helper.au3
Comment (mlipok) [reply wd_helper.au3]: Done

**Comment by mlipok** [comment]
I notice that:
`_WD_UpdateDriver('msedge', @ScriptDir , False, True)`
or
`_WD_UpdateDriver('msedge', @ScriptDir , True, True)`

extracts also `Driver_Notes` directory .
Is it needed ?

File: wd_helper.au3
Comment (mlipok) [reply wd_helper.au3:1477]: > I'm ok with changing it to "0", "0.00", or whatever.

Let stick with 0



> We just need to clearly document the change as potentially script breaking just in case someone is checking for "none".

in REVISIONS.md ?

Comment (Danp2) [reply wd_helper.au3:1477]: Yes. I have already started the revised list locally, so I'll add it there.

**Review by Danp2 — CHANGES_REQUESTED** [review]
Should be "0" for consistency with other function

**Comment by Danp2** [comment]
> extracts also Driver_Notes directory . Is it needed ?

That comes from the MSEdge webdriver. It isn't needed, but would require changing the extraction code to only copy the desired executable.

**Comment by mlipok** [comment]
> Should be "0" for consistency with other function.
> 
> Edit: Now I see that you changed the other one back to "None". Please explain.

Currentyl there is not used "None" 
It is set as `Local $sDriverVersion = 0` 

Do you mean that this should be `Local $sDriverVersion = "0"` ?

**Comment by mlipok** [comment]
> It isn't needed, but would require changing the extraction code to only copy the desired executable.

I will make a relevant fix that extracts only EXE files.

**Comment by Danp2** [comment]
> Currentyl there is not used "None"

I may be mistaken, but I thought that you switched _WD_GetWebDriverVersion back to "None"

> Do you mean that this should be Local $sDriverVersion = "0" ?

Yes. It's the same thing we discussed yesterday where you are assigning a numeric value to a string variable. 

IMO, both _WD_GetWebDriverVersion and _WD_GetBrowserVersion should return the same result for failure (ie: both return "None" or both return "0"). Pick one and stick to it. 😉

**Comment by mlipok** [comment]
I choose
> both return "0"

because `"None"` dose'nt work with `_VersionCompare()`

**Comment by mlipok** [comment]
> I choose
> 
> > both return "0"
> 
> because `"None"` doesn't work with `_VersionCompare()`

done

File: wd_helper.au3
Comment (mlipok) [reply wd_helper.au3]: Done

**Review by Danp2 — APPROVED** [review]

**Review by Danp2 — APPROVED** [review]

**Pull request merged**
